### PR TITLE
fix(careful): gate destructive guards on execution context and boundaries

### LINKED ISSUE #835: fix(careful): scope destructive SQL detection to executable context
## Problem

`hooks/careful-check.sh` scans the entire Bash command string for destructive SQL phrases. Even after requiring `TRUNCATE` statement shape, read-only commands still produce confirmation prompts:

```text
grep truncate somefile        -> Destructive SQL prompt
rg "DROP TABLE" src          -> Destructive SQL prompt
printf "TRUNCATE users"       -> Destructive SQL prompt
```

This is a Claude Octopus hook false positive, not a host/harness error. It interrupts governed and unattended runs for harmless source searches and output commands.

## Durable acceptance criteria

- Read-only search/output contexts such as `grep truncate somefile`, `rg "DROP TABLE" src`, and `printf "TRUNCATE users"` stay quiet.
- Known SQL-client execution (`psql -c "TRUNCATE users"`, `mysql -e "DROP TABLE users"`) still asks.
- Direct destructive SQL statements still ask.
- Piped SQL into a known client still asks.
- Behavioral regression tests invoke the real hook with careful mode active.

PR #831 fixes related substring/boundary cases and is the intended implementation vehicle for this context-aware hardening.

### PR DESCRIPTION
**Pull request by klass-723**
## Problem

`hooks/careful-check.sh` matched destructive-looking substrings anywhere in a Bash command. Harmless source searches such as the reported Tailwind `truncate` grep, `grep truncate somefile`, and `rg "DROP TABLE" src` could interrupt governed or unattended runs with a destructive-SQL confirmation prompt. The sibling `rm` and whole-tree checkout rules also matched `charm -rf`, dotfiles, and `./src`.

## Fix

- SQL checks now require both statement shape and executable context: a direct destructive statement or a known SQL client (`psql`, `mysql`, `mariadb`, `sqlite3`, `sqlcmd`, `cockroach sql`).
- Read-only grep/rg/printf examples stay quiet; SQL client flags and piped SQL still ask.
- `rm` requires a command boundary.
- `git checkout/restore .` only matches the whole-tree path, not dotfiles or `./src`.
- Behavioral regressions invoke the real hook with careful mode active.

## Verification

- `bash tests/smoke/test-safety-hooks.sh`: 26/26 passing
- `bash -n hooks/careful-check.sh`
- `git diff --check`
- Remote Linux/macOS/unit/symlink matrix runs on each update.

Fixes #835.

Original diagnosis and boundary fix by @klass-723; execution-context hardening added during maintainer review.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved careful-mode detection for destructive commands.
  * Reduced false positives when matching `rm`, `TRUNCATE`, and `git checkout` or `git restore`.
  * Commands are now recognized using complete command and argument patterns, avoiding matches within unrelated text or filenames.
  * Preserved existing protections for destructive database, filesystem, and Git operations.
  * Added regression coverage for command-boundary and statement-shape detection.

* **Documentation**
  * Updated the changelog with details about the strengthened command detection.

### CODE REVIEW

File: hooks/careful-check.sh [hunk 1]
```diff
@@ -107,8 +113,12 @@ if echo "$CHECK_TEXT" | grep -qE 'git\s+reset\s+--hard'; then
...
 fi
 
-# 5. git checkout . / git restore .
-if echo "$CHECK_TEXT" | grep -qE 'git\s+checkout\s+\.|git\s+restore\s+\.'; then
+# 5. git checkout . / git restore . — only the WHOLE-TREE discard (`.` or bare `./`)
+# should warn "discards all unstaged changes". Bare `\.` also matched a single dotfile
+# (`git checkout .gitignore`), and `\.(/)` also matched a `./`-prefixed single path
+# (`git checkout ./.gitignore`, `git restore ./.env`) — both discard one file, not all.
+# So the dot must be followed by an optional slash and then end/space, nothing else.
+if echo "$CHECK_TEXT" | grep -qE 'git\s+(checkout|restore)\s+\.(/)?(\s|$)'; then
```
[the remark this answers is not in the document]
Comment (nyldn) [reply hooks/careful-check.sh]: Fixed in ca60ec21: the whole-tree matcher now accepts whitespace, semicolon, pipe, and ampersand separators while keeping dotfiles and ./subpaths quiet. The behavioral suite covers chained checkout and restore forms.

File: tests/smoke/test-safety-hooks.sh [hunk 2]
```diff
@@ -168,6 +168,49 @@ test_careful_returns_ask_decision() {
...
+    [[ "$(_cc_decides 'git checkout ./.gitignore')" == quiet ]] || fails+=" checkout-slash-dotfile"
+    [[ "$(_cc_decides 'git restore ./.env')"       == quiet ]] || fails+=" restore-slash-dotfile"
+    [[ "$(_cc_decides 'git checkout ./src')"       == quiet ]] || fails+=" checkout-subpath"
+    # Real destructive ops that must still FIRE (no false negative introduced).
+    [[ "$(_cc_decides 'psql -c "TRUNCATE users"')" == fire ]] || fails+=" truncate-sql"
+    [[ "$(_cc_decides 'TRUNCATE TABLE foo')"       == fire ]] || fails+=" truncate-table"
+    [[ "$(_cc_decides 'rm -rf /tmp/somewhere')"    == fire ]] || fails+=" rm-rf"
+    [[ "$(_cc_decides 'foo; rm -rf /etc')"         == fire ]] || fails+=" rm-rf-chained"
+    [[ "$(_cc_decides 'git checkout .')"           == fire ]] || fails+=" checkout-dot"
+    [[ "$(_cc_decides 'git checkout ./')"          == fire ]] || fails+=" checkout-dotslash"
```
[the remark this answers is not in the document]
Comment (nyldn) [reply tests/smoke/test-safety-hooks.sh:219]: Fixed in ca60ec21: added positive cases for git restore ., git restore ./, and chained checkout/restore variants. Safety hook suite passes 26/26.

File: hooks/careful-check.sh [hunk 3]
```diff
@@ -85,9 +85,18 @@ if echo "$CHECK_TEXT" | grep -qE 'rm\s+-[a-zA-Z]*r[a-zA-Z]*f|rm\s+-r\s+-f|rm\s+-
...
+# searches and output commands routinely contain examples such as `DROP TABLE` and
+# `TRUNCATE users`. Gate only when the statement is either a direct shell command or
+# appears alongside a known SQL client. This keeps read-only grep/rg/printf commands
+# quiet while retaining coverage for client flags, stdin pipes, and heredocs.
+_octo_sql_pat='DROP\s+TABLE|DROP\s+DATABASE|TRUNCATE\s+(TABLE\s+)?["'\''`]?\w'
+_octo_sql_client_pat='(^|[^[:alnum:]_])(psql|mysql|mariadb|sqlite3|sqlcmd|cockroach\s+sql)([^[:alnum:]_]|$)'
+_octo_direct_sql_pat="^[[:space:]]*(${_octo_sql_pat})"
+if echo "$CHECK_TEXT" | grep -qiE "$_octo_sql_pat" \
+    && { echo "$CHECK_TEXT" | grep -qiE "$_octo_sql_client_pat" \
+        || echo "$CHECK_TEXT" | grep -qiE "$_octo_direct_sql_pat"; }; then
```
[the remark this answers is not in the document]
Comment (nyldn) [reply hooks/careful-check.sh]: Fixed in d4c8d853. TRUNCATE candidates now get token validation so TABLE requires a third target token; direct TRUNCATE TABLE and psql -c "TRUNCATE TABLE;" stay quiet, while TRUNCATE users and TRUNCATE TABLE foo still fire. Safety suite passes 26/26 and ShellCheck is clean.

File: hooks/careful-check.sh [hunk 4]
```diff
@@ -107,8 +116,12 @@ if echo "$CHECK_TEXT" | grep -qE 'git\s+reset\s+--hard'; then
...
 fi
 
-# 5. git checkout . / git restore .
-if echo "$CHECK_TEXT" | grep -qE 'git\s+checkout\s+\.|git\s+restore\s+\.'; then
+# 5. git checkout . / git restore . — only the WHOLE-TREE discard (`.` or bare `./`)
+# should warn "discards all unstaged changes". Bare `\.` also matched a single dotfile
+# (`git checkout .gitignore`), and `\.(/)` also matched a `./`-prefixed single path
+# (`git checkout ./.gitignore`, `git restore ./.env`) — both discard one file, not all.
+# So the dot must be followed by an optional slash and then end/space, nothing else.
+if echo "$CHECK_TEXT" | grep -qE 'git\s+(checkout|restore)\s+\.(/)?([[:space:];|&]|$)'; then
```
[the remark this answers is not in the document]
Comment (nyldn) [reply hooks/careful-check.sh]: Fixed in d527a028. The whole-tree matcher now covers quoted paths and optional -- while keeping quoted ./subpaths quiet. Added positive cases for checkout/restore quoted and -- forms plus a negative quoted-subpath case; safety suite passes 26/26 and ShellCheck is clean.

File: hooks/careful-check.sh [hunk 5]
```diff
@@ -85,9 +85,42 @@ if echo "$CHECK_TEXT" | grep -qE 'rm\s+-[a-zA-Z]*r[a-zA-Z]*f|rm\s+-r\s+-f|rm\s+-
...
-if echo "$CHECK_TEXT" | grep -qiE 'DROP\s+TABLE|DROP\s+DATABASE|TRUNCATE'; then
-    matched=$(echo "$CHECK_TEXT" | grep -oiE 'DROP\s+TABLE|DROP\s+DATABASE|TRUNCATE' | head -1)
+# 2. SQL destructive operations. A SQL-looking string alone is not execution: source
+# searches and output commands routinely contain examples such as `DROP TABLE` and
+# `TRUNCATE users`. Gate only when the statement is either a direct shell command or
+# appears alongside a known SQL client. This keeps read-only grep/rg/printf commands
+# quiet while retaining coverage for client flags, stdin pipes, and heredocs.
+_octo_drop_pat='DROP\s+TABLE|DROP\s+DATABASE'
+_octo_truncate_pat='TRUNCATE\s+["'\''`]?[[:alnum:]_]+(\s+["'\''`]?[[:alnum:]_]+)?'
+_octo_sql_pat="${_octo_drop_pat}|${_octo_truncate_pat}"
```
[the remark this answers is not in the document]
Comment (nyldn) [reply hooks/careful-check.sh]: Fixed in af480d92. DROP TABLE/DATABASE candidates now require a target, preserve IF EXISTS, and reject incomplete forms before execution-context gating. Added quiet incomplete/client cases and positive psql/mysql target cases. Real-hook probes match the expected decisions; safety 26/26 and ShellCheck clean.
